external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -3028,7 +3028,9 @@ fn resolve_index_funs_and_type(
             .add_diag(diag!(Declarations::MissingSyntaxMethod, (loc, msg()),));
         return (None, context.error_type(loc));
     };
-    let fty = core::make_function_type(context, loc, &m, &f, None);
+    // NOTE: We don't do a visibility check here because we _just_ care about computing the return
+    // type. The visibility check will happen later in `exp_to_borrow_`.
+    let fty = core::make_function_type_no_visibility_check(context, loc, &m, &f, None);
```

Comment:
What is `f` here? Or rather, what does `index.get_name_for_typing()` resolve to?

Reply:
The `index.get_name_for_typing` method will return the imm. ref. `index` if it is defined, or the mut. ref. `index_mut` if `index` is not. Since all other arguments are the same, this is sufficient to compute the base return type, which is essential for computing out more of the path.

We can't check visibility here due to the following example:

```
    #[syntax(index)]
    fun index_b(b: &B): &B { b }

    #[syntax(index)]
    public fun index_b_mut(b: &mut B): &mut B { b }
 ```

If we call `&mut b[]` in a different module, checking visibility here will report a "phantom" visibility error. Instead, we resolve simply to figure out the return type, then do proper function resolution/visibility checking in `exp_to_borrow_` (as the comment suggests). 

I had an alternative draft that attempted to restrict visibility when retrieving the syntax definition, but the errors are much less helpful: the definition is simply "not found." If we wanted better errors, I suppose we could do that and _also_ make the fields in the `IndexSyntaxMethods` struct into an enum:

```
enum SyntaxMethodEntry {
  VisibilityError(...),
  SyntaxMethod(Box<SyntaxMethod>),
  Undefined
}
```

This seemed a bit convoluted to me, but I'm happy to do whichever you think is best.

Reply:
I was imagining that this choice of letting the syntax methods have different visibility would just let the chips fall where they may on expansion. Really thats my goal here, just do the thing that happens naturally. 

I guess in that case, we should force the two methods to have the same visibility? Thoughts?